File: cheesy_cover_letter.py
'''
Determines whether or not to interview a Candidate, "me". Assumes equal weight of all assessment criteria (25%).
**Needs refactoring:
    1) Need to add sectional breaks for each raw_input, such that it only asks the incorrectly
    answered question again instead of asking all questions again
'''
# dict for converting "low-high" values to numbers
val_pairs = {"low": 1,"med-low": 2,"med": 3,"med-high": 4,"high": 5}

# define Candidate class
class Candidate(object):
    def __init__(self):
        # 'pro' attributes
        self.motivation = "high"
        self.fast_learner = "high"
        self.has_potential = "med-high"
        # 'con' attributes
        self.formal_training = "low"
        self.experience = "med-low"

    def to_number(self,a):
        return val_pairs[self.__dict__[a]]

# define functions to check inputs
def prob_found():
    print " -- ENTER APPROPRIATE INPUT, PLEASE! -- "
    print " -- I NEED TO ASK EVERYTHING AGAIN -- "
    print
    return False

def check_number_input(input):
    if (len(input) > 1) or (len(input) == 0) or (1 > int(input)) or (int(input) > 5):
##            print "ed = ", ed
##            print "len of ed > 1: ", (len(ed) > 1)
##            print "len of ed = 0: ", (len(ed) == 0)
##            print "1 > ed: ", 1 > int(ed)
##            print "ed > 5: ", int(ed) > 5
##            candidate_can_debug = True
        return prob_found()        

# define function for calculating if candidate should get interview   
def worth_interview(candidate):
    education = ""
    exp_or_pot = ""
    motivation = ""
    ability_to_train = ""
    
    while True:
        # prompt/calculate importance of education
        education = raw_input("Does a formal education matter significantly for this position(y/n)? ")
        if education == "n":
            education = 15 + candidate.to_number("formal_training")*2
        elif education == "y":
            education = candidate.to_number("formal_training")*5
        else:
            prob_found()
            break
        print "---"

        # prompt/calculate preference: experience or potential       
        exp_or_pot = raw_input("Which is more important for this position: experience or potential('e' or 'p')? ")
        if (exp_or_pot == "e"):
            exp_or_pot = 5*candidate.to_number('experience')
        elif (exp_or_pot == "p"):
            exp_or_pot = 5*candidate.to_number('has_potential')
        else:
            prob_found()
            break
        print "---"

        # prompt/calculate importance of motivation
        motivation = raw_input("On a scale of 1 to 5 (5 being the most), how important is the candidate's motivation? ")
        if check_number_input(motivation) == False:
            break
        motivation = int(motivation)*candidate.to_number('motivation')
        print "---"

        # prompt/calculate time to train potential candidates
        ability_to_train = raw_input("On a scale of 1 to 5 (5 being the most), how much time can you afford to spend training the candidate, if hired? ")
        if check_number_input(ability_to_train) == False:
            break
        ability_to_train = int(ability_to_train)*candidate.to_number('fast_learner')
        print "---"

        # calculate candidate score
        cand_score = 0
        for i in [education, exp_or_pot,motivation,ability_to_train]:
            cand_score += i
        break
    
    # start over if any of the inputs were invalid    
    if "" in [education, exp_or_pot, motivation, ability_to_train]:
        return worth_interview(candidate)

    print "Candidate score is: ", cand_score
    if cand_score > 60:
        print "Definitely worth a shot!"
    else:
        print "It's a stretch, but you never know!"
           
me = Candidate()   
worth_interview(me)

File: cheesy_cover_letter.rb
# By: Thomas Brigham
# Phone: 571-435-5250
# Email: tkbrigham@gmail.com
# Github: https://github.com/tkbrigham

####
# Determines whether or not to interview a Candidate, "me". Assumes equal 
# weight of all assessment criteria (25%). 
####


######
### DEFINE CANDIDATE CLASS
######
class Candidate
	@@rating_to_val = { low: 1, med_low: 2, med: 3, med_high: 5, high: 5 }
	
	attr_accessor :motivation, :fast_learner, :has_potential, :formal_training, :experience

	def initialize(motivation, fast_learner, has_potential, formal_training, experience)
		# 'pro' attirbutes
		@motivation = motivation
		@fast_learner = fast_learner
		@has_potential = has_potential
		# 'con' attributes
		@formal_training = formal_training
		@experience = experience
	end

	def convert_to_val(att)
		@@rating_to_val[self.method(att).call.to_sym].to_i
	end

	def worth_interview?
		education = determine_education(self)
		puts "-------"
		
		exp_or_pot = determine_exp_or_pot(self)
		puts "-------"
		
		motivation = determine_motivation(self)
		puts "-------"
		
		training = determine_training(self)
		puts "-------"

		cand_score = calculate_score(education, exp_or_pot, motivation, training)
		
		puts "Candidate score is: #{cand_score}"
		if cand_score > 60
			puts "Definitely worth a shot!"
		else
			puts "It's a stretch, but you never know!"
		end
	end
end

######
### PROGRAM FUNCTIONS
######
def prob_found
	puts
	puts " -- ENTER APPROPRIATE INPUT, PLEASE! -- "
	puts
end

def check_number_input(input)
	if 1 > input.to_i || input.to_i > 5
		return false
	end
end

# prompt/calculate importance of education
def determine_education(candidate)
	puts "Does a formal education matter significantly for this position(y/n)? "
	education = gets.chomp
	if education == "n"
		education = 15 + candidate.convert_to_val("formal_training")*2
	elsif education == "y"
		education = candidate.convert_to_val("formal_training")*5
	else
		prob_found
		determine_education(candidate)
	end		
end

# prompt/calculate preference: experience or potential
def determine_exp_or_pot(candidate)
	puts "Which is more important for this position: experience or potential('e' or 'p')? "
	exp_or_pot = gets.chomp
	if exp_or_pot == "e"
		exp_or_pot = 5*candidate.convert_to_val('experience')
	elsif exp_or_pot == "p"
		exp_or_pot = 5*candidate.convert_to_val('has_potential')
	else
		prob_found
		determine_exp_or_pot(candidate)
	end	
end

# prompt/caluclate importance of motivation
def determine_motivation(candidate)
	puts "On a scale of 1 to 5 (5 being the most), how important is the candidate's motivation?"
	motivation = gets.chomp
	if check_number_input(motivation) == false
		prob_found
		determine_motivation(candidate)
	else
		motivation = motivation.to_i*candidate.convert_to_val('fast_learner')
	end
end

# prompt/caluclate ability to train/mentor potential candidates
def determine_training(candidate)
	puts "On a scale of 1 to 5 (5 being the most), how much training and/or mentoring are you planning on providing to the candidate?"
	ability_to_train = gets.chomp
	if check_number_input(ability_to_train) ==false
		prob_found
		determine_training(candidate)
	else
		ability_to_train = ability_to_train.to_i*candidate.convert_to_val('fast_learner')
	end
end

# calculate candidate score
def calculate_score(education, exp_or_pot, motivation, training)
	cand_score = [education, exp_or_pot, motivation, training].reduce(:+)
end

me = Candidate.new(
	motivation = "high",
	fast_learner = "high", 
	has_potential = "med_high", 
	formal_training = "low", 
	experience = "med")

me.worth_interview?

